### ISSUE TITLE
Hypertuning parameters

### ISSUE THREAD

**Issue by malangadan** [malangadan/Challengers#9]
Currently every model is doing pretty well, but they could do better! Please look at the documentation for each model or review the slides to understand what tuning each hyper parameter does so you can explain it in the presentation. Ideally we want to see some increase in f1 score and other metrics. 

Use something like [grid search](https://sckit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html) it will test many combinations of hyperparameters. I will post some reference code you can use to get idea of what you should do.

Models like knn should see some tangible increase from this for sure.

**Comment by malangadan** [comment]
from sklearn.model_selection import GridSearchCV

knn = KNeighborsClassifier()

param_grid = {
    'n_neighbors': [3,4,5,6],
    'weights': ['uniform', 'distance'],
    'algorithm': ['auto','brute'],
}

grid_search = GridSearchCV(estimator=knn, param_grid=param_grid, cv=5, n_jobs=-1)

grid_search.fit(X_train, y_train)

print("Best Hyperparameters:", grid_search.best_params_)

best_knn = grid_search.best_estimator_

y_pred = best_knn.predict(X_test)

accuracy = best_knn.score(X_test, y_test)
print(accuracy)

precision = precision_score(y_test, y_pred, average='binary')
print(f"Precision: {precision}")

recall = recall_score(y_test, y_pred, average='binary')
print(f"Recall: {recall}")

f1 = f1_score(y_test, y_pred, average='binary')
print(f"F1 Score: {f1}")

cm = confusion_matrix(y_test, y_pred)

plt.figure(figsize=(6, 4))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', xticklabels=['Class 0', 'Class 1'], yticklabels=['Class 0', 'Class 1'])
plt.title('Confusion Matrix')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.show()
